capella_console_client/repeat_requests.py

```diff
@@ -0,0 +1,145 @@
+    ObservationDirection,
+    OrbitState,
+    ProductType,
+    ProductClass,
+    OrbitalPlane,
+    RepeatCollectionTier,
+    InstrumentMode,
+    Polarization,
+    ArchiveHoldback,
+)
+
+COLLECT_CONSTRAINTS_KEYS = set(
+    [
+        "collect_mode",
+        "look_direction",
+        "asc_dsc",
+        "orbital_planes",
+        "local_time",
+        "off_nadir_min",
+        "off_nadir_max",
+        "elevation_min",
+        "elevation_max",
+        "image_length",
+        "image_width",
+        "azimuth",
+        "grr_min",
+        "grr_max",
+        "srr_min",
+        "srr_max",
+        "azr_min",
+        "azr_max",
+        "nesz_max",
+        "num_looks",
+        "polarization",
+    ]
+)
+
+REPETITION_PROPERTIES_KEYS = set(
+    [
+        "repeat_start",
+        "repeat_end",
+        "repetition_interval",
+        "repetition_count",
+        "maintain_scene_framing",
+        "look_angle_tolerance",
+    ]
+)
+
+
+def create_repeat_request(
+    session: CapellaConsoleSession,
+    geometry: geojson.geometry.Geometry,
+    name: Optional[str] = "",
+    description: Optional[str] = "",
+    collection_tier: Optional[Union[str, RepeatCollectionTier]] = RepeatCollectionTier.routine,
+    product_category: Optional[Union[str, ProductClass]] = ProductClass.standard,
+    archive_holdback: Optional[Union[str, ArchiveHoldback]] = ArchiveHoldback.none,
+    custom_attribute_1: Optional[str] = None,
+    custom_attribute_2: Optional[str] = None,
+    product_types: Optional[List[Union[str, ProductType]]] = None,
+    # Repetition properties
+    repeat_start: Optional[Union[datetime, str]] = None,
+    repeat_end: Optional[Union[datetime, str]] = None,
+    repetition_interval: Optional[int] = 7,
+    repetition_count: Optional[int] = None,
+    maintain_scene_framing: Optional[bool] = False,
+    look_angle_tolerance: Optional[int] = None,
+    # Collect constraints
+    collect_mode: Optional[Union[str, InstrumentMode]] = InstrumentMode.spotlight,
+    look_direction: Optional[Union[str, ObservationDirection]] = ObservationDirection.either,
+    asc_dsc: Optional[Union[str, OrbitState]] = OrbitState.either,
+    orbital_planes: Optional[List[Union[int, OrbitalPlane]]] = None,
+    local_time: Optional[List[int]] = None,
+    off_nadir_min: Optional[int] = None,
+    off_nadir_max: Optional[int] = None,
+    elevation_min: Optional[int] = None,
+    elevation_max: Optional[int] = None,
+    image_length: Optional[int] = None,
+    image_width: Optional[int] = None,
+    azimuth: Optional[int] = None,
+    grr_min: Optional[int] = None,
+    grr_max: Optional[int] = None,
+    srr_min: Optional[int] = None,
+    srr_max: Optional[int] = None,
+    azr_min: Optional[int] = None,
+    azr_max: Optional[int] = None,
+    nesz_max: Optional[int] = None,
+    num_looks: Optional[int] = None,
+    polarization: Optional[Union[str, Polarization]] = None,
+) -> Dict[str, Any]:
+
+    if isinstance(repeat_start, str):
```

Comment:
can you extract this into something like

```python
def datetime_to_iso8601_str(default_value: datetime, dt: Optional[Union[datetime, str]] = None) -> str:
    if not dt:
        dt = default_value
    else:
        try:
            dt = parse(dt)
        except ParserError:
            raise ValueError(f"Could not parse {dt} string into a datetime")
    return dt.isoformat(timespec="milliseconds") + "Z"
```

put it in `validate.py` and changes lines 101 to 117 to use that? Same goes for `tasking.py`

Reply:
Very nice, much cleaner. Had to slightly modify it but changed all the instances to use that function.